components/gitpod-protocol/src/context-url.ts

```diff
@@ -4,41 +4,88 @@
  * See License-AGPL.txt in the project root for license information.
  */
 
+import { Workspace } from ".";
+
 export namespace ContextURL {
   export const INCREMENTAL_PREBUILD_PREFIX = "incremental-prebuild";
   export const PREBUILD_PREFIX = "prebuild";
   export const IMAGEBUILD_PREFIX = "imagebuild";
+  export const SNAPSHOT_PREFIX = "snapshot";
   export const REFERRER_PREFIX = 'referrer:';
 
+  export function getNormalized(ws: Pick<Workspace, "context" | "contextURL"> | undefined, fallbackToRaw: boolean = true): string | undefined {
+    const normalized = normalize(ws);
+    if (!normalized && fallbackToRaw) {
+      return ws?.contextURL;
+    }
+    return normalized;
+  }
+
+  export function getNormalizedURL(ws: Pick<Workspace, "context" | "contextURL">): URL | undefined {
+    const contextURL = normalize(ws);
+    if (!contextURL) {
+      return undefined;
+    }
+
+    try {
+      return new URL(contextURL);
+    } catch (err) {
+      console.error(err);
+    }
+    return undefined;
+  }
+
+  function normalize(ws: Pick<Workspace, "context" | "contextURL"> | undefined): string | undefined {
+    if (!ws) {
+      return undefined;
+    }
+    if (ws.context.normalizedContextURL) {
+      return ws.context.normalizedContextURL;
+    }
+
+    // fallback: we do not yet set normalizedContextURL on all workspaces, yet, let alone older existing workspaces
+    try {
+      const u = removePrefixes(ws.contextURL);
+      if (u) {
+        return u;
+      }
+    } catch (err) {
+      console.error(err);
+    }
+
+    return undefined;
+  }
```

Comment:
I agree that in most cases we should prefer `ws.context.normalizedContextURL`. However, isn't it a little overkill to have two different methods, `getNormalized` and `getNormalizedURL`?

I think I would prefer having just `getNormalizedURL` to keep this protocol helper as simple as possible. _(Side note: I find the function name `getNormalized(workspace)` a little bit ambiguous, and had to read the code to understand what it does.)_

Reply:
Same answer as here: https://github.com/gitpod-io/gitpod/pull/8099#discussion_r801868556